File: backend/user-service/src/main/java/com/banking/userservice/service/IdentityVerificationService.java
package com.banking.userservice.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

@Service
public class IdentityVerificationService {

    private static final Logger logger = LoggerFactory.getLogger(IdentityVerificationService.class);

    @Autowired
    private RestTemplate restTemplate;

    @Value("${SOCURE_API_KEY:mock-socure-key}")
    private String socureApiKey;

    @Value("${JUMIO_API_TOKEN:mock-jumio-token}")
    private String jumioApiToken;

    @Value("${HIBP_API_KEY:mock-hibp-key}")
    private String hibpApiKey;

    public void verifyIdentityAsync(String email, Long userId) {
        logger.info("Starting identity verification for user {} ({})", userId, email);

        CompletableFuture<Void> socure = CompletableFuture.runAsync(() -> callSocure(email))
                .orTimeout(5, TimeUnit.SECONDS)
                .exceptionally(ex -> { logger.warn("Socure check failed: {}", ex.getMessage()); return null; });

        CompletableFuture<Void> jumio = CompletableFuture.runAsync(() -> callJumio(email, userId))
                .orTimeout(5, TimeUnit.SECONDS)
                .exceptionally(ex -> { logger.warn("Jumio check failed: {}", ex.getMessage()); return null; });

        CompletableFuture<Void> hibp = CompletableFuture.runAsync(() -> callHibp(email))
                .orTimeout(5, TimeUnit.SECONDS)
                .exceptionally(ex -> { logger.warn("HIBP check failed: {}", ex.getMessage()); return null; });

        CompletableFuture.allOf(socure, jumio, hibp)
                .thenRun(() -> logger.info("Identity verification complete for user {}", userId));
    }

    private void callSocure(String email) {
        logger.info("Calling Socure ID+ for email: {}", email);
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("Authorization", "SocureApiKey " + socureApiKey);

        Map<String, Object> body = Map.of(
                "modules", new String[]{"emailrisk"},
                "email", email
        );

        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(body, headers);
        restTemplate.postForEntity("https://sandbox.socure.com/api/3.0/EmailAuthScore", entity, String.class);
        logger.info("Socure ID+ call completed for email: {}", email);
    }

    private void callJumio(String email, Long userId) {
        logger.info("Calling Jumio for user: {}", userId);
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("Authorization", "Bearer " + jumioApiToken);

        Map<String, Object> body = Map.of(
                "customerInternalReference", "apex-user-" + userId,
                "userReference", email
        );

        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(body, headers);
        restTemplate.postForEntity("https://netverify.com/api/v4/initiate", entity, String.class);
        logger.info("Jumio call completed for user: {}", userId);
    }

    private void callHibp(String email) {
        logger.info("Calling HIBP for email: {}", email);
        HttpHeaders headers = new HttpHeaders();
        headers.set("hibp-api-key", hibpApiKey);
        headers.set("user-agent", "ApexBanking/1.0");

        HttpEntity<Void> entity = new HttpEntity<>(headers);
        restTemplate.exchange(
                "https://haveibeenpwned.com/api/v3/breachedaccount/" + email,
                HttpMethod.GET,
                entity,
                String.class
        );
        logger.info("HIBP call completed for email: {}", email);
    }
}

File: backend/user-service/Services/IdentityVerificationService.cs
using System.Text;
using System.Text.Json;

namespace UserServiceDotnet.Services;

public class IdentityVerificationService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<IdentityVerificationService> _logger;
    private readonly string _socureApiKey;
    private readonly string _jumioApiToken;
    private readonly string _hibpApiKey;

    public IdentityVerificationService(
        HttpClient httpClient,
        ILogger<IdentityVerificationService> logger,
        IConfiguration configuration)
    {
        _httpClient = httpClient;
        _logger = logger;
        _socureApiKey = configuration["SOCURE_API_KEY"] ?? "mock-socure-key";
        _jumioApiToken = configuration["JUMIO_API_TOKEN"] ?? "mock-jumio-token";
        _hibpApiKey = configuration["HIBP_API_KEY"] ?? "mock-hibp-key";
    }

    public void VerifyIdentityAsync(string email, long userId)
    {
        _logger.LogInformation("Starting identity verification for user {UserId} ({Email})", userId, email);

        _ = Task.Run(async () =>
        {
            var socure = CallSocure(email);
            var jumio = CallJumio(email, userId);
            var hibp = CallHibp(email);

            try
            {
                await Task.WhenAll(socure, jumio, hibp);
            }
            catch
            {
                // individual tasks already log their own failures
            }

            _logger.LogInformation("Identity verification complete for user {UserId}", userId);
        });
    }

    private async Task CallSocure(string email)
    {
        try
        {
            _logger.LogInformation("Calling Socure ID+ for email: {Email}", email);

            using var request = new HttpRequestMessage(HttpMethod.Post, "https://sandbox.socure.com/api/3.0/EmailAuthScore");
            request.Headers.Add("Authorization", $"SocureApiKey {_socureApiKey}");
            request.Content = new StringContent(
                JsonSerializer.Serialize(new { modules = new[] { "emailrisk" }, email }),
                Encoding.UTF8, "application/json");

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
            await _httpClient.SendAsync(request, cts.Token);
            _logger.LogInformation("Socure ID+ call completed for email: {Email}", email);
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Socure check failed: {Message}", ex.Message);
        }
    }

    private async Task CallJumio(string email, long userId)
    {
        try
        {
            _logger.LogInformation("Calling Jumio for user: {UserId}", userId);

            using var request = new HttpRequestMessage(HttpMethod.Post, "https://netverify.com/api/v4/initiate");
            request.Headers.Add("Authorization", $"Bearer {_jumioApiToken}");
            request.Content = new StringContent(
                JsonSerializer.Serialize(new { customerInternalReference = $"apex-user-{userId}", userReference = email }),
                Encoding.UTF8, "application/json");

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
            await _httpClient.SendAsync(request, cts.Token);
            _logger.LogInformation("Jumio call completed for user: {UserId}", userId);
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Jumio check failed: {Message}", ex.Message);
        }
    }

    private async Task CallHibp(string email)
    {
        try
        {
            _logger.LogInformation("Calling HIBP for email: {Email}", email);

            using var request = new HttpRequestMessage(HttpMethod.Get,
                $"https://haveibeenpwned.com/api/v3/breachedaccount/{email}");
            request.Headers.Add("hibp-api-key", _hibpApiKey);
            request.Headers.Add("user-agent", "ApexBanking/1.0");

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
            await _httpClient.SendAsync(request, cts.Token);
            _logger.LogInformation("HIBP call completed for email: {Email}", email);
        }
        catch (Exception ex)
        {
            _logger.LogWarning("HIBP check failed: {Message}", ex.Message);
        }
    }
}
